dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/ProgrammaticFaceletTests.java

```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.org.apache.myfaces40.fat.tests;
+
+import java.util.logging.Logger;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+
+import componenttest.annotation.Server;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.HttpUtils;
+
+/**
+ * Programmatic Facelet Tests for Faces 4.0.
+ *
+ * https://github.com/jakartaee/faces/issues/1581
+ *
+ */
+@RunWith(FATRunner.class)
+public class ProgrammaticFaceletTests {
+
+    private static final Logger LOG = Logger.getLogger(SimpleTest.class.getName());
+    private static final String PROGRAMMATIC_FACELET_TEST_APP_NAME = "ProgrammaticFaceletTests";
+
+    @Server("faces40_programmaticFaceletServer")
+    public static LibertyServer server;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        ShrinkHelper.defaultDropinApp(server, PROGRAMMATIC_FACELET_TEST_APP_NAME + ".war",
+                                      "io.openliberty.org.apache.faces40.fat.programmaticfacelettests");
+
+        server.startServer(SimpleTest.class.getSimpleName() + ".log");
+    }
+
+    @AfterClass
+    public static void testCleanup() throws Exception {
+        LOG.info("testCleanUp : stop server");
+
+        // Stop the server
+        if (server != null && server.isStarted()) {
+            server.stopServer();
+        }
+    }
+
+    /**
+     * Basic test of the use of the @View annotation.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testBasicView() throws Exception {
+        HttpUtils.findStringInReadyUrl(server, "/" + PROGRAMMATIC_FACELET_TEST_APP_NAME + "/BasicView.xhtml", "Hello World!");
```

Comment:
I don't know too much about HttpUtils. If the text is not found, is the page logged?   Could logging be added?

Reply:
Yes, the page is logged as follows:
`The response did not contain "[Hello World!ABC]". Full output is:" &lt;html xmlns=&quot;http://www.w3.org/1999/xhtml&quot; xmlns:h=&quot;http://xmlns.jcp.org/jsf/html&quot;&gt;<head></head><body><form id="testForm" name="testForm" method="post" action="/ProgrammaticFaceletTests/BasicView.xhtml" enctype="application/x-www-form-urlencoded">Hello World!<input type="hidden" name="testForm_SUBMIT" value="1" /><input type="hidden" name="jakarta.faces.ViewState" id="j_id__v_0:jakarta.faces.ViewState:1" value="ZTAwN2U2YWIzZTU1ZWEwZDAwMDAwMDAx" autocomplete="off" /></form></body>&lt;/html&gt; ". expected:<0> but was:<1>
`